### ISSUE TITLE
Exception thrown in network interceptor hangs coroutine

### ISSUE THREAD

**Issue by robfletcher** [square/retrofit#3110]
If an `okhttp3.Interceptor` throws an exception, the `suspend` function in a Retrofit interface never completes (either returning a value or throwing an exception).

A sample project demonstrating the issue with a simple JUnit test can be found here: robfletcher/retrofit-coroutines-hang

The real situation when I've encountered this is when an SSL handshake fails.

FWIW the same behavior is seen using JakeWharton/retrofit2-kotlin-coroutines-adapter with Retrofit 2.5.0

**Comment by swankjesse** [comment]
When an SSL handshake fails that should throw an IOException, which should be recovered gracefully.

This is an IllegalStateException, which typically results in an application crash.

**Comment by robfletcher** [comment]
It does the same thing if the interceptor throws `IOException`.

**Comment by JakeWharton** [comment]
This should be trivial to test. It's likely we already have coverage for
IOException subtypes but probably worth doing both.

**Comment by JakeWharton** [comment]
This test shows `IOException` being propagated properly

https://github.com/square/retrofit/blob/master/retrofit/src/test/java/retrofit2/KotlinSuspendTest.kt#L85-L99

**Comment by robfletcher** [comment]
Sorry, away from computer right now but does the exception in that case get raised from an interceptor? I’ve seen exceptions propagate successfully in other cases. The issue I’m hitting seems to specifically pertain to interceptors.

I’ll update my example when I get to work in a couple hours.

**Comment by JakeWharton** [comment]
OkHttp is implemented as a chain of interceptors. The last interceptor just
talks to a Socket instead of calling chain.proceed. So yes, but beyond
that, Retrofit doesn't care about from where the exception originates in
OkHttp since it only talks to OkHttp through one API:
Call.enqueue(Callback).

**Comment by robfletcher** [comment]
Ok, my mistake. I could swear I had tested this with `IOException` but when I updated my example it no longer fails. We must be doing something that ends up throwing another type of exception in our SSL interceptor (integrating with Netflix's metatron cert auth system).

**Comment by JakeWharton** [comment]
Still probably worth a test

**Comment by JakeWharton** [comment]
So this is a severe design flaw in OkHttp it turns out. Non-`IOException`s thrown in an `enqueue`'d request result in no callback whatsoever.

**Comment by robfletcher** [comment]
Glad I wasn't wasting your time!

**Comment by JakeWharton** [comment]
Blocked by https://github.com/square/okhttp/issues/5151

**Comment by CharlyLafon37** [comment]
Hi, are there any updates on this issue ? There's a workaround where we can catch any exception thrown in Interceptors and rethrow them as IOException but it would be nice to have a fix soon :)

**Comment by Hodkinson** [comment]
Just adding a workaround here in case it helps. Actually in my case even IOExceptions are not propagating back. Retrofit 2.6.2, okhttp 4.2.0

I wanted to deal with this at point of call rather than add an UncaughtExceptionHandler.

So the workaround was to add an interceptor which catches any IOExceptions, and returns a blank response with a magic status code.

```
private val errorInterceptor = object : Interceptor {
        override fun intercept(chain: Interceptor.Chain): okhttp3.Response {
            return try {
                chain.proceed(chain.request())
            } catch (e: IOException) {
                okhttp3.Response.Builder()
                    .code(418)
                    .request(chain.request())
                    .body(object: ResponseBody() {
                        override fun contentLength() = 0L
                        override fun contentType(): MediaType? = null
                        override fun source(): BufferedSource = Buffer()
                    })
                    .message(e.message ?: e.toString())
                    .protocol(Protocol.HTTP_1_1)
                    .build()
            }
        }
    }
```

This could be combined with a Result object like in the answer here: https://stackoverflow.com/questions/57625272/retrofit-2-6-0-custom-coroutines-calladapterfactory where the status code or message is checked and a NetworkError returned.

**Comment by Nghicv** [comment]
Hi, are there any updates on this issue?
